### ISSUE TITLE
How to use both port and securePort in CephObjectStore when both are specified

### ISSUE THREAD

**Issue by ushitora-anqou** [rook/rook#12614]
**Is this a bug report or feature request?**
* Bug Report

Currently, I provide non-secure RGW endpoints and am trying to support TLS endpoints too. It's better for the existing users to continue to use the non-secure endpoints without any modification.

To achieve the aforementioned goal, I specified both port and securePort (*1). However, I encountered the following problems.

a. How to use both non-secure ports and TLS ports is not documented.
b. If securePort is set, the `BUCKET_PORT` environment variables in ConfigMaps corresponding to the existing OBCs, are changed to use a TLS endpoint.

Problem b is caused by [this code.](https://github.com/rook/rook/blob/cb44c7b88fb346f3b120e8bda769ac808ec60880/pkg/apis/ceph.rook.io/v1/object.go#L49) I found that non-secure endpoints can be used by creating new StorageClasses having endpoint field. However, I hesitate to adopt this approach because of the following reasons.

- The endpoint field [seems to be for external clusters](https://github.com/rook/rook/commit/b86bb9c5286d7bc953848907be0dee7bccb3c27c).
- I don't want to write the endpoint's IP address directly in StorageClass.

(*1) According to [the ObjectStore design docs](https://github.com/rook/rook/blob/5979b3816292df6b3ff17eb8e58eb8cf88488802/design/ceph/object/store.md#gateway), both port and securePort can be set.

**Environment**:
* Rook version (use `rook version` inside of a Rook Pod): v1.11.6
* Storage backend version (e.g. for ceph do `ceph -v`): v17.2.6

**Comment by thotz** [comment]
Documentation can updated, please feel free to do so since you have more idea about what is missing.

For the endpoint related info, Incase of OBC, only one port and endpoint can be shared. Thats limitation of OBC library and since it is deprecated we cannot make that change. But we can have similar changes for CephObjectStoreUser and request feature in upcoming COSI to include multiple endpoints

**Comment by ushitora-anqou** [comment]
I'll update the document, but I would prefer to do that only after my problem has been solved. I want to let users use either secure or insecure endpoints depending on their purpose. I understood it's not accomplished by OBC. Then, is there any way to set the `BUCKET_PORT` environment variable depending on users without using the endpoint field of StorageClass?

I'd like not to use the endpoint field if possible. It's because the original purpose of the endpoint field is for external clusters, and writing endpoint information to StorageClass is redundant.

**Comment by thotz** [comment]
The `BUCKET_PORT` is populated by lib-bucket-provisioner based on the port value which is given by Rook Operator. If the `endpoint` is not defined in the storage class, Rook Operator picks up the port from the CephObjectStore Spec. Since we can only pass one port Operator will pick up the secure one than insecure port.
To resolve this we can either make settings in Storageclass level, like `secure` `false` or in ObjectStoreSpec for all access resources . The first one is more granular same object store can provide OBCs which can use secure or insecure port. But changes to depreciated lib bucket provisioner does not make sense at the moment. Second option we are to limit access to one specific port for all the resources like OBC, COSI Bucket/BucketAccess and CephObjectStoreUser consumed by the cephobjectstore

**Comment by ushitora-anqou** [comment]
How about supporting a new field like 'secure: false' in additionalConfig than in the underneath of spec. Since additionalConfig is storage provider-specific, we don't need to change lib-bucket-provisioner itself. Actually, using additionalConfig was [suggested to add the way to set sharding configuration](https://github.com/rook/rook/issues/12100#issuecomment-1517424128) although this idea wasn't adopted at that time.

In addition, although users will continue to use lib-bucket-provisioner for several years since the COSI driver is in an experimental state.

**Comment by thotz** [comment]
Yaa that also works, My main concern is regarding changes for OBC or StorageClass in Rook Operator, The idea to depreciate OBC 1.13 or 1.14 depends on the stability of COSI(so max 2 years IMO)

**Comment by ushitora-anqou** [comment]
> My main concern is regarding changes for OBC or StorageClass in Rook Operator,

Does this mean that you consider that all changes to OBC or StorageClass in the Rook operator are not allowed anymore?
If not, My team member or I will create a PR that implements my idea.

**Comment by satoru-takeuchi** [comment]
@travisn 

> Does this mean that you consider that all changes to OBC or StorageClass in the Rook operator are not allowed anymore?
If not, My team member or I will create a PR that implements my idea.

Do you have any thoughts? If this way is not acceptable, if users specify both `port` and `securePort`, they can't use the non-secure endpoint without knowing the value of `port` beforehand. Since specifying both ports has been valid for a long time, the change that @ushitora-anqou proposed is not to support a new feature but a bug fix.

**Comment by travisn** [comment]
To be clear, OBCs and lib-bucket-provisioner are not yet deprecated, but development effort is just very limited while we are focusing new development on COSI. So if someone submits PRs related to OBCs for bug fixes, we should still take them. 

I'm not understand the fix needed for using both ports, so we can follow up on the details when a PR is opened.

**Comment by ushitora-anqou** [comment]
I will create a new pull request if needed. I will close this issue. Thank you for your responses and comments!
